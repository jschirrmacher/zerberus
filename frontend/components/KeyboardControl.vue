<script setup lang="ts">
const emit = defineEmits(["turn", "move", "zoom"])

const commands = {
  ArrowDown: () => emit("move", { x: 0, y: 50 }),
  ArrowUp: () => emit("move", { x: 0, y: -50 }),
  ArrowLeft: () => emit("move", { x: -50, y: 0 }),
  ArrowRight: () => emit("move", { x: 50, y: 0 }),
  ShiftArrowDown: () => emit("turn", { x: -5, y: 0 }),
  ShiftArrowUp: () => emit("turn", { x: 5, y: 0 }),
  ShiftArrowLeft: () => emit("turn", { x: 0, y: 5 }),
  ShiftArrowRight: () => emit("turn", { x: 0, y: -5 }),
  Plus: () => emit("zoom", 10),
  Minus: () => emit("zoom", -10),
} as Record<string, () => void>

const keyMap = {
  "+": "Plus",
  "-": "Minus",
}

function keydown(event: KeyboardEvent) {
  const key = keyMap[event.key as keyof typeof keyMap] || event.key
  const code = (event.shiftKey ? "Shift" : "") + key
  commands[code] && commands[code]()
}

window.addEventListener("keydown", keydown)
</script>

<template>
  <nav>
    <button id="ArrowUp" @click="commands.ArrowUp">ᐃ</button>
    <button id="ArrowLeft" @click="commands.ArrowLeft">ᐊ</button>
    <button id="ArrowRight" @click="commands.ArrowRight">ᐅ</button>
    <button id="ArrowDown" @click="commands.ArrowDown">ᐁ</button>
    <button id="Plus" @click="commands.Plus">+</button>
    <button id="Minus" @click="commands.Minus">&nbsp;</button>
  </nav>
</template>

<style scoped lang="scss">
button {
  height: 26px;
  width: 26px;
  line-height: 0px;
}

#Minus::before {
  content: "–";
  position: absolute;
  left: 7px;
  top: 10px;
}

#Plus::after,
#Minus::after {
  content: "";
  border: 1px solid #000000;
  width: 14.5px;
  height: 14.5px;
  display: inline-block;
  position: absolute;
  border-radius: 50%;
  top: 3.5px;
  left: 3.5px;
}
</style>
