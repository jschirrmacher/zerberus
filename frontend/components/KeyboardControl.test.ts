import { describe, it, expect } from "vitest"
import { shallowMount, VueWrapper } from "@vue/test-utils"
import KeyboardControl from "./KeyboardControl.vue"

type AugmentedVM = VueWrapper & { keydown: (obj: object) => void }

describe("KeyboardControl", () => {
  it("should render as expected", () => {
    const component = shallowMount(KeyboardControl)
    expect(component.element).toMatchSnapshot()
  })

  it("should emit move events when a button is clicked", () => {
    const component = shallowMount(KeyboardControl)
    component.find("#ArrowUp").trigger("click")
    expect(component.emitted().move).toContainEqual([{ x: 0, y: -50 }])
  })

  it("should emit move events when an arrow key is pressed", () => {
    const component = shallowMount(KeyboardControl)
    ;(component.vm as unknown as AugmentedVM).keydown({ key: "ArrowDown" })
    expect(component.emitted().move).toContainEqual([{ x: 0, y: 50 }])
  })

  it("should emit turn events when an arrow key and the shift key is pressed", () => {
    const component = shallowMount(KeyboardControl)
    ;(component.vm as unknown as AugmentedVM).keydown({ key: "ArrowDown", shiftKey: true })
    expect(component.emitted().turn).toContainEqual([{ x: -5, y: 0 }])
  })

  it("should emit zoom events when the plus key is pressed", () => {
    const component = shallowMount(KeyboardControl)
    ;(component.vm as unknown as AugmentedVM).keydown({ key: "+" })
    expect(component.emitted().zoom).toContainEqual([10])
  })

  it("should emit zoom events when the minus key is pressed", () => {
    const component = shallowMount(KeyboardControl)
    ;(component.vm as unknown as AugmentedVM).keydown({ key: "-" })
    expect(component.emitted().zoom).toContainEqual([-10])
  })
})
