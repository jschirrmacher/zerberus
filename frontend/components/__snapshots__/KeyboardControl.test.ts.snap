// Vitest Snapshot v1

exports[`KeyboardControl > should render as expected 1`] = `
<nav
  data-v-89b5344a=""
>
  <button
    data-v-89b5344a=""
    id="ArrowUp"
  >
    ᐃ
  </button>
  <button
    data-v-89b5344a=""
    id="ArrowLeft"
  >
    ᐊ
  </button>
  <button
    data-v-89b5344a=""
    id="ArrowRight"
  >
    ᐅ
  </button>
  <button
    data-v-89b5344a=""
    id="ArrowDown"
  >
    ᐁ
  </button>
  <button
    data-v-89b5344a=""
    id="Plus"
  >
    +
  </button>
  <button
    data-v-89b5344a=""
    id="Minus"
  >
     
  </button>
</nav>
`;
